pkg/globalnet/controllers/ipam/gatewaymonitor.go

```diff
@@ -16,63 +16,74 @@ limitations under the License.
+	"github.com/submariner-io/admiral/pkg/watcher"
+	v1 "github.com/submariner-io/submariner/pkg/apis/submariner.io/v1"
 	"github.com/submariner-io/submariner/pkg/cidr"
 	"github.com/submariner-io/submariner/pkg/globalnet/cleanup"
 	"github.com/submariner-io/submariner/pkg/iptables"
 	"github.com/submariner-io/submariner/pkg/netlink"
-
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
-	"k8s.io/apimachinery/pkg/util/wait"
-	"k8s.io/client-go/informers"
-	"k8s.io/client-go/kubernetes"
-	"k8s.io/client-go/tools/cache"
-	"k8s.io/client-go/util/workqueue"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes/scheme"
 	"k8s.io/klog"
-
-	v1 "github.com/submariner-io/submariner/pkg/apis/submariner.io/v1"
-	submarinerClientset "github.com/submariner-io/submariner/pkg/client/clientset/versioned"
-	submarinerInformers "github.com/submariner-io/submariner/pkg/client/informers/externalversions"
 )
 
-const defaultResync = 60 * time.Second
-
-func NewGatewayMonitor(spec *SubmarinerIPAMControllerSpecification,
-	submarinerClient submarinerClientset.Interface, clientSet kubernetes.Interface,
-	dynamicClient dynamic.Interface) (*GatewayMonitor, error) {
+func NewGatewayMonitor(spec *SubmarinerIPAMControllerSpecification, config watcher.Config) (*GatewayMonitor, error) {
 	gatewayMonitor := &GatewayMonitor{
-		clusterID:         spec.ClusterID,
-		endpointWorkqueue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "Endpoints"),
-		submarinerInformerFactory: submarinerInformers.NewSharedInformerFactoryWithOptions(submarinerClient,
-			time.Second*30, submarinerInformers.WithNamespace(spec.Namespace)),
-		submarinerClientSet: submarinerClient,
-		kubeClientSet:       clientSet,
-		dynamicClientSet:    dynamicClient,
-		ipamSpec:            spec,
-		isGatewayNode:       false,
+		clusterID:     spec.ClusterID,
+		ipamSpec:      spec,
+		isGatewayNode: false,
 	}
 
-	endpointInformer := gatewayMonitor.submarinerInformerFactory.Submariner().V1().Endpoints()
-
 	iptableHandler, err := iptables.New()
 	if err != nil {
 		return nil, err
 	}
 
 	gatewayMonitor.ipt = iptableHandler
-	gatewayMonitor.endpointsSynced = endpointInformer.Informer().HasSynced
+
+	if config.RestMapper == nil {
+		if config.RestMapper, err = admUtil.BuildRestMapper(config.RestConfig); err != nil {
+			return nil, err
+		}
+	}
+
+	if config.Client == nil {
+		if config.Client, err = dynamic.NewForConfig(config.RestConfig); err != nil {
+			return nil, fmt.Errorf("error creating dynamic client: %v", err)
+		}
+	}
+
+	if config.Scheme == nil {
+		config.Scheme = scheme.Scheme
+	}
+
+	config.ResourceConfigs = []watcher.ResourceConfig{
+		{
+			Name:         "IPAM GatewayMonitor",
+			ResourceType: &v1.Endpoint{},
+			Handler: watcher.EventHandlerFuncs{
+				OnCreateFunc: gatewayMonitor.handleCreatedOrUpdatedEndpoint,
```

Comment:
What do you think about moving right away to the https://github.com/submariner-io/submariner/blob/devel/pkg/event/handler.go interface instead.

The event handler also provides updates on the nodes. As well as transitions to gateways or non-gateways.

@sridhargaddam  @tpantelis ^ ?

Reply:
yeah we can look into using that interface but I'd rather do that in another PR. We need to move this forward to proceed with GN 2.0 enhamcements.

Reply:
@tpantelis that sounds good this PR already implements a good deal of improvements and testing. It should be easier to move from this one if we finally see that as reasonable.